### PR TITLE
Add subset="both" functionality to {image|text}_dataset_from_directory()

### PR DESCRIPTION
**Pull request by Haaris-Rahman**
Add subset="both" functionality to text_dataset_from_directory()

### CODE REVIEW

File: keras/preprocessing/text_dataset.py [hunk 1]
```diff
@@ -153,30 +153,66 @@ def text_dataset_from_directory(directory,
...
-      # Shuffle locally at each iteration
-      dataset = dataset.shuffle(buffer_size=batch_size * 8, seed=seed)
-    dataset = dataset.batch(batch_size)
+  if subset == "both":
+    file_paths_train, labels_train = dataset_utils.get_training_or_validation_split(
+        file_paths, labels, validation_split, subset)
+    file_paths_val, labels_val = dataset_utils.get_training_or_validation_split(
+        file_paths, labels, validation_split, subset)
+    if not file_paths_train or file_paths_val:
+      raise ValueError(f'No text files found in directory {directory}. '
```
Comment (fchollet) [inline keras/preprocessing/text_dataset.py:162]: What if only one of the two datasets is empty? In what situation could that occur? Could we provide a better message in that case?

File: keras/preprocessing/text_dataset.py [hunk 2]
```diff
@@ -153,30 +153,66 @@ def text_dataset_from_directory(directory,
...
+                      f'Allowed format: .txt')
+    dataset_train = paths_and_labels_to_dataset(
```
Comment (fchollet) [inline keras/preprocessing/text_dataset.py:164]: Use `train_dataset` and `val_dataset` as the names

File: keras/preprocessing/text_dataset.py [hunk 3]
```diff
@@ -153,30 +153,66 @@ def text_dataset_from_directory(directory,
...
+        num_classes=len(class_names),
+        max_length=max_length)
+    
+    dataset_train = dataset_train.prefetch(tf.data.AUTOTUNE)
+    dataset_val = dataset_val.prefetch(tf.data.AUTOTUNE)
+    if batch_size is not None:
+      if shuffle:
+        # Shuffle locally at each iteration
+        dataset_train = dataset_train.shuffle(buffer_size=batch_size * 8, seed=seed)
+        dataset_val = dataset_val.shuffle(buffer_size=batch_size * 8, seed=seed)
```
Comment (fchollet) [inline keras/preprocessing/text_dataset.py:183]: Validation data normally does not shuffling. Did we previously shuffle the validation dataset?

File: keras/preprocessing/text_dataset.py [hunk 4]
```diff
@@ -153,30 +153,66 @@ def text_dataset_from_directory(directory,
...
+      if shuffle:
+        # Shuffle locally at each iteration
+        dataset = dataset.shuffle(buffer_size=batch_size * 8, seed=seed)
+      dataset = dataset.batch(batch_size)
+    else:
+      if shuffle:
+        dataset = dataset.shuffle(buffer_size=1024, seed=seed)
+
+    # Users may need to reference `class_names`.
+    dataset.class_names = class_names
```
Comment (fchollet) [inline keras/preprocessing/text_dataset.py:215]: This bit of metadata should also be added to each dataset in the "both" case.

File: keras/preprocessing/text_dataset.py [hunk 5]
```diff
@@ -153,30 +153,66 @@ def text_dataset_from_directory(directory,
...
+    dataset_val = dataset_val.prefetch(tf.data.AUTOTUNE)
+    if batch_size is not None:
+      if shuffle:
+        # Shuffle locally at each iteration
+        dataset_train = dataset_train.shuffle(buffer_size=batch_size * 8, seed=seed)
+        dataset_val = dataset_val.shuffle(buffer_size=batch_size * 8, seed=seed)
+      dataset_train = dataset_train.batch(batch_size)
+      dataset_val = dataset_val.batch(batch_size)
+    else:
+      if shuffle:
```
Comment (fchollet) [inline keras/preprocessing/text_dataset.py:187]: It seems the `if shuffle` block is repeated twice, you can move it out of the conditional block

File: keras/preprocessing/text_dataset.py [hunk 6]
```diff
@@ -153,30 +153,66 @@ def text_dataset_from_directory(directory,
...
+        num_classes=len(class_names),
+        max_length=max_length)
+    dataset = dataset.prefetch(tf.data.AUTOTUNE)
+    if batch_size is not None:
+      if shuffle:
+        # Shuffle locally at each iteration
+        dataset = dataset.shuffle(buffer_size=batch_size * 8, seed=seed)
+      dataset = dataset.batch(batch_size)
+    else:
+      if shuffle:
```
Comment (fchollet) [inline keras/preprocessing/text_dataset.py:211]: Same here

**Review by fchollet — CHANGES_REQUESTED** [review]
Thanks for the PR! 👍

File: keras/preprocessing/text_dataset.py (hunk 5 above)
Comment (Haaris-Rahman) [reply keras/preprocessing/text_dataset.py:187]: I believe this is required. One if batch size is mentioned and another if batch size is not mentioned.

File: keras/preprocessing/text_dataset.py (hunk 1 above)
Comment (Haaris-Rahman) [reply keras/preprocessing/text_dataset.py:162]: I believe it would only occur when validation_split is zero or one. Will fix it

File: keras/preprocessing/text_dataset_test.py [hunk 7]
```diff
@@ -174,6 +174,16 @@ def test_text_dataset_from_directory_validation_split(self):
     batch = next(iter(dataset))
     self.assertLen(batch, 2)
     self.assertEqual(batch[0].shape, (2,))
+    
+    train_dataset, val_dataset = text_dataset.text_dataset_from_directory(
```
Comment (fchollet) [inline keras/preprocessing/text_dataset_test.py:178]: Please also add this test for `image_dataset_from_directory`.

**Review by fchollet — COMMENTED** [review]
Thanks for the updates!

File: keras/preprocessing/text_dataset_test.py (hunk 7 above)
Comment (fchollet) [reply keras/preprocessing/text_dataset_test.py:178]: I missed it -- I see it's added already.

**Review by fchollet — APPROVED** [review]
LGTM, thanks!

**Comment by Haaris-Rahman** [comment]
```
Error in download_and_extract: java.io.IOException: Error downloading [https://zlib.net/zlib-1.2.11.tar.gz] to /home/kbuilder/.cache/bazel/_bazel_kbuilder/31d6f47147b75c35404d734345be7323/external/zlib/temp2526511131029126787/zlib-1.2.11.tar.gz: GET returned 404 Not Found
```
I faced a similar issue with bezel test. For some reason zlib-1.2.11.tar.gz does not exist. I used zlib-1.2.12.tar.gz .on my local machine.

**Comment by fchollet** [comment]
> I faced a similar issue with bezel test. For some reason zlib-1.2.11.tar.gz does not exist. I used zlib-1.2.12.tar.gz .on my local machine.

This is linked to a security issue that was discovered in zlib the other day. Don't worry about it, it's not related to the contents of your PR. I'll investigate how to merge on Monday.

**Comment by fchollet** [comment]
An update: we're bumping the zlib version number on our side. When that change is merged, we'll be able to merge this PR.

**Comment by haifeng-jin** [comment]
A rebase may be needed to pass CI tests?

**Review by fchollet — APPROVED** [review]

**Comment by fchollet** [comment]
@Haaris-Rahman can you please rebase the PR on master?

**Comment by Haaris-Rahman** [comment]
> @Haaris-Rahman can you please rebase the PR on master?

Yes, working on it.
